backend/pcloud/chunkwriter.go

```diff
@@ -0,0 +1,337 @@
+		client:    client,
+		fs:        srcFs,
+		chunkSize: info.ChunkSize,
+		src:       src,
+		remote:    remote,
+		fd:        openResult.FileDescriptor,
+		fileID:    openResult.Fileid,
+	}
+
+	return info, writer, nil
+}
+
+// Abort implements fs.ChunkWriter.
+func (c *chunkWriter) Abort(ctx context.Context) error {
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open aborted: %w", err)
+	}
+	if err := o.Remove(ctx); err != nil {
+		return fmt.Errorf("delete aborted: %w", err)
+	}
+	return nil
+}
+
+// Close implements fs.ChunkWriter.
+func (c *chunkWriter) Close(ctx context.Context) error {
+	// In case the file existed previosly and was larger than the newly written file, we need to truncate it to the right size.
+	byteCount := c.GetBytesCount()
+	fs.Debugf(c.src, "truncate target to %d bytes", byteCount)
+	if _, err := c.fileTruncate(ctx, byteCount); err != nil {
+		return fmt.Errorf("truncate file: %w", err)
+	}
+
+	// close fd
+	if _, err := c.fileClose(ctx); err != nil {
+		return fmt.Errorf("close fd: %w", err)
+	}
+
+	// There seems to be no option in the fileops API to set the modtime.
+	// Therefore, we apply the modtime via the regular file API.
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open newly copied object: %w", err)
+	}
+	if err := o.SetModTime(ctx, c.src.ModTime(ctx)); err != nil {
+		return fmt.Errorf("set modtime: %w", err)
+	}
+
+	return nil
+}
+
+// WriteChunk implements fs.ChunkWriter.
+func (c *chunkWriter) WriteChunk(ctx context.Context, chunkNumber int, reader io.ReadSeeker) (bytesWritten int64, err error) {
+	return c.writeChunk(ctx, chunkNumber, reader)
+}
+
+func (c *chunkWriter) writeChunk(ctx context.Context, chunkNumber int, reader io.Reader) (bytesWritten int64, err error) {
+	chunk := int64(chunkNumber)
+
+	// read chunk into buffer
+	buffer := make([]byte, c.chunkSize)
+	offset := chunk * c.chunkSize
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	// read input buffer and calculate hash
+	inSHA1 := ""
+	eg.Go(func() error {
+		byteCount, err := io.ReadFull(reader, buffer)
+		buffer = buffer[:byteCount]
+		if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) && !errors.Is(err, io.EOF) {
+			return err
+		}
+		inSHA1Bytes := sha1.Sum(buffer)
+		inSHA1 = hex.EncodeToString(inSHA1Bytes[:])
+		return nil
+	})
+
+	// get target hash
+	outSHA1 := ""
+	eg.Go(func() error {
+		targetChecksum, err := c.fileChecksum(egCtx, offset, c.chunkSize)
+		if err != nil {
+			return err
+		}
+		outSHA1 = targetChecksum.SHA1
+		return nil
+	})
+
+	if errors.Is(err, io.EOF) {
+		return 0, nil
+	}
+
+	if err := eg.Wait(); err != nil {
+		return 0, fmt.Errorf("check hashes of chunk %d: %w", chunk, err)
+	}
+	if outSHA1 == "" || inSHA1 == "" {
+		return 0, fmt.Errorf("block %d: expect both hashes to be filled: src: %q, target: %q", chunk, inSHA1, outSHA1)
+	}
+
+	byteCount := int64(len(buffer))
+	c.AddBytesCount(byteCount)
+	// check hash of chunk, skip if fits
```

Comment:
This check is checking to see if existing chunks in the file match to see if we can skip the upload.

If we turn on the `PartialUploads` flag then this test will no longer work.

Also if you uploading a big file that has lots of changes (say a compressed `.tar.gz`) then this will effectively download the file and upload it.

So I'm not terribly keen on this test anyway.

Reply:
This check is the essential part of the chunked upload in my use-case:

I use rclone with pcloud to backup many large media files from local.
As uploading of large files often fails with pcloud (e.g. by a server side ConnReset), so I need to have a way to resume failed uploads of large files.

The hash for pcloud isn't a problem here, as they luckily offer an API for partial hashes.
While I understand that uploading a large file from another non-local remote may cause downloading the whole file just for checking the issues: The download would happen anyway, as the upload is runs through the local RAM as well, while this check would at least save the upload (which is probably the critical thing in most cases).

We could however skip this check by checking the file size, e.g. by storing the originial file size in `Open` and skipping this check when offset+chunksize runs out of bounds.

Just skipping this check wouldn't be an option for me. Alternatively, is there another concept in rclone for resuming failed uploads?

What do you think?